Fix EagerLoader loading optional association with leftJoinWith and contain

victoreassis opened the pull request:
Solves #14281 

It seems to work.

I added a test on ORM/QueryTest.php to check:
- If it works when there's some results with the associated filled and some not (this already worked as intended).
- If it works when all the results have a null associate (this was the case that needed a fix).
- If the table doesn't have an association (the exception that was thrown before).

I really don't know if it is on the right place and if it is a good enough test.

victoreassis committed:
Fix EagerLoader to avoid exception on loading optional association with leftJoinWith and contain

victoreassis committed:
Add test case on loading optional association with leftJoinWith and contain

victoreassis committed:
Fix code style errors

victoreassis committed:
Fix EagerLoader to still be able to throw the no association exception if the needed field was not selected on the query

markstory commented on src/ORM/EagerLoader.php:
```diff
@@ -830,13 +830,18 @@ protected function _groupKeys(BufferedStatement $statement, array $collectKeys):
+                        continue;
+                    }
+                    // Assign empty array to avoid not found association when optional.
+                    if (empty($keys[$nestKey][$parts[0]]) && !isset($result[$parts[1][0]])) {
```
Should this be `isset()` instead of empty()? empty() can swallow '' and 0.

victoreassis replied:
I guess it could be. I will change it right away, although I guess it will never have those values as the way the array is built, but is better to be safe.

victoreassis replied:
I did the change and some other one to avoid the presence of ```['' => null]``` on the result keys set when there's some null associations.

markstory commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -3412,6 +3412,105 @@ public function testLeftJoinWithAutoFields()
+                'published' => 'N',
+                'author' => null,
+            ],
+        ];
```
What is this test covering that the first insert/read isn't? They look very similar to me.

victoreassis replied:
The first test covers when there's result with and without the association, some associations are null, and the second covers when there's no results with association, all associations are null.

victoreassis replied:
I made some changes to tests to remove the second insert doing a query only on null associations.

markstory reviewed (commented):
Thanks for taking the initiative to put this together :clap:

victoreassis committed:
Change empty to !isset call to check if there are any keys present on _groupKeys method

Add condition to avoid indexed array with empty string when key is null but the result keys set is already set.

victoreassis committed:
Change test case removing second insert replacing by a where clause on the second assert

markstory commented:
Thank you 👍
